java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -86,7 +88,38 @@
     private final KinesisProducerConfiguration config;
     private final Map<String, String> env;
     private final AtomicLong messageNumber = new AtomicLong(1);
-    private final Map<Long, SettableFuture<?>> futures = new ConcurrentHashMap<>();
+    private final Map<Long, SettableFutureTracker> futures = new ConcurrentHashMap<>();
+    private final PriorityQueue<SettableFutureTracker> oldestFutureTrackerHeap = new PriorityQueue(new SettableFutureTrackerComparator());
+
+    private class SettableFutureTrackerComparator implements Comparator<SettableFutureTracker>
+    {
+        public int compare(SettableFutureTracker x, SettableFutureTracker y)
+        {
+            if(y.getTimestamp().getEpochSecond() > x.getTimestamp().getEpochSecond()) {
+                return 1;
+            }
+            return -1;
+        }
+    }
+
+
+    private class SettableFutureTracker {
+        private SettableFuture<?> future;
+        private Instant timestamp;
+
+        public SettableFutureTracker(SettableFuture<?> f, Instant timestamp) {
```

Comment:
We need non-null checks here to avoid NPEs in comparator. Introducing lombok will make life easier as you can get replace of all these constructors, setters, getters, null checks with annotations.

Reply:
Done